Prefabs/Interactions/Controllables/SharedResources/Scripts/DirectionalDrive.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace VRTK.Core.Prefabs.Interactions.Controllables
+{
+    using UnityEngine;
+    using VRTK.Core.Data.Type;
+
+    /// <summary>
+    /// The basis to drive a control in a linear direction.
+    /// </summary>
+    public abstract class DirectionalDrive : Drive
+    {
+        #region Facade Settings
+        /// <summary>
+        /// The public interface facade.
+        /// </summary>
+        [Header("Facade Settings"), Tooltip("The public interface facade.")]
+        public DirectionalDriveFacade facade;
+        #endregion
+
+        /// <summary>
+        /// Calculates the limits of the drive.
+        /// </summary>
+        /// <param name="newLimit">The maximum local space limit the drive can reach.</param>
+        /// <returns>The minimum and maximum local space limit the drive can reach.</returns>
+        public abstract FloatRange CalculateDriveLimits(float newLimit);
+
+        /// <inheritdoc />
+        protected override void InternalSetup()
+        {
+            baseFacade = facade;
```

Comment:
The Property for Facade is a good call.

I don't quite understand the making it a Generic though.

If I did something like

`abstract class Drive<T> : Monobehaviour`

then in the subclass did

`class DirectionalDrive : Drive<DirectionalDrive>`

Then the `Drive` class won't be able to access any of the methods on T (and the Drive class calls methods and properties in the abstract class now)

It may be my limited knowledge of generics is causing my lack of understanding.

(also the `InternalSetup` method is used later on in the RotationalJointDrive to set up joint stuff so it can't be removed, but the later comment about the verb name is legit)

Reply:
```c#
public abstract class Drive<T> : MonoBehaviour where T : DriveFacade
{
    public T facade;
}

public class DirectionalDrive : Drive<DirectionalDrive>
{
}
```

That should work and actually means there is no need for a property as the superclass can now just define the facade **field** that will be serialized properly as the subclasses "fix" the generic type which is what makes Unity happy.

👍 on the `InternalSetup` method staying then (after the rename). At least it means besides the joint stuff there will be no need to do anything in most classes with the above generic field suggestion.